chore: update contributing guide with new process for unconventional dependencies (#22851)
aws/aws-cdk: CONTRIBUTING.md

* Make sure to update the PR title/description if things change. The PR title/description are going to be used as the
  commit title/message and will appear in the CHANGELOG, so maintain them all the way throughout the process.

#### Adding dependencies bundled in lambda layers

Sometimes, constructs introduce runtime dependencies bundled in lambda layers for use in custom resources.
The CDK includes three of these as direct dependencies for legacy purposes: `@aws-cdk/asset-awscli-v1`,
`@aws-cdk/asset-kubectl-v20`, and `@aws-cdk/asset-node-proxy-agent-v5`. The actual dependencies are pulled out
to separate repositories (like this [one](https://github.com/cdklabs/awscdk-asset-kubectl) for kubectl)
and managed there. If you would like to introduce additional runtime dependencies, it likely involves discussing

> with a CDK maintainer and opening a new repository in cdklabs that vends the lambda layer. Generally, each branch

Reviewer: [nit] Do we have a public list that we could point to?

Reviewer 2: We don't have one, and it would be too annoying to maintain :/ I do agree, it would be nice though.

on the repository will focus on a specific version of the dependency. For example, in `awscdk-asset-kubectl`,
branch `kubectl-v20/main` vends kubectl v1.20, branch `kubectl-v21/main` vends kubectl v1.21, and so on.

After creating the repository, there are two possible paths to follow on a case-by-case basis:

- Add a direct dependency in the module's `package.json` on a specific version of the dependency. This means that you
  lock yourself to a specific version of the dependency, which makes sense for something like `awscli` v1.

> - In addition to having a direct dependency acting as a default, expose a property on the construct you are creating

Reviewer: [nit] This is not a different option, it's an increment of the step above -- right?
Otherwise "in addition" reads strangely in the paragraph

Reviewer 2: +1, I think this is addressed in my review.

  that allows users to supply their own version of the dependency if need be. This makes the most sense for dependencies
  like `kubectl`, which have multiple minor versions to choose from.
  
To see an example of this in action, take a look at how this works in the `eks.Cluster`

> [construct](https://github.com/aws/aws-cdk/blob/main/packages/@aws-cdk/aws-eks/README.md#L685-L692).

Reviewer: ```suggestion
[construct](https://github.com/aws/aws-cdk/blob/main/packages/@aws-cdk/aws-eks/README.md?plain=1#L685-L692).
```

> **If your PR introduces runtime dependencies in lambda layers, make sure to call it out int he description

Reviewer: ```suggestion
**If your PR introduces runtime dependencies in lambda layers, make sure to call it out in the description
```
